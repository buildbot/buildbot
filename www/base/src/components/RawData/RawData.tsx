/*
  This file is part of Buildbot.  Buildbot is free software: you can
  redistribute it and/or modify it under the terms of the GNU General Public
  License as published by the Free Software Foundation, version 2.

  This program is distributed in the hope that it will be useful, but WITHOUT
  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
  FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
  details.

  You should have received a copy of the GNU General Public License along with
  this program; if not, write to the Free Software Foundation, Inc., 51
  Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

  Copyright Buildbot Team Members
*/

import './RawData.scss';
import moment from 'moment';
import {useState} from 'react';
import _ from 'underscore';
import {isObservableArray, isObservableObject} from 'mobx';
import {ArrowExpander} from 'buildbot-ui';

const isArrayRaw = (v: any) => {
  return _.isArray(v) || isObservableArray(v);
};
const isObjectRaw = (v: any) => {
  return (_.isObject(v) || isObservableObject(v)) && !isArrayRaw(v);
};

const isArrayOfObjectsRaw = (v: any) => {
  return isArrayRaw(v) && v.length > 0 && isObjectRaw(v[0]);
};

export const displayBuildRequestEntry = (key: string, value: any): string => {
  if (key === 'claimed_at' || key === 'complete_at' || key === 'submitted_at') {
    try {
      return `${value} (${moment(Number.parseInt(value.toString()) * 1000).format()})`;
    } catch (error) {
      return value.toString();
    }
  }
  return value.toString();
};

export const displayBuildsetEntry = (key: string, value: any): string => {
  if (key === 'complete_at' || key === 'created_at' || key === 'submitted_at') {
    // created_at is for dictionaries in sourcestamps list
    try {
      return `${value} (${moment(Number.parseInt(value.toString()) * 1000).format()})`;
    } catch (error) {
      return value.toString();
    }
  }
  return value.toString();
};

type RawDataProps = {
  data: {[key: string]: any};
  displayCallback?: (key: string, value: any) => string;
};

export const RawData = ({data, displayCallback}: RawDataProps) => {
  const [isExpanded, setIsExpanded] = useState(false);

  const renderArrayElements = (value: any[]) => {
    return (
      <ul>
        {value.map((v, index) => (
          <li key={index}>
            <RawData data={v} displayCallback={displayCallback} />
          </li>
        ))}
      </ul>
    );
  };

  const renderDataElement = (key: string, value: any) => {
    if (value === null) {
      return <dd>{'null'}&nbsp;</dd>;
    }
    if (value === undefined) {
      return <dd>{'undefined'}&nbsp;</dd>;
    }
    if (isArrayOfObjectsRaw(value)) {
      return (
        <dd>
          <ArrowExpander isExpanded={isExpanded} setIsExpanded={setIsExpanded} />
          {isExpanded ? renderArrayElements(value as any[]) : <span>{JSON.stringify(value)}</span>}
        </dd>
      );
    }
    if (isObjectRaw(value)) {
      return (
        <dd>
          <ArrowExpander isExpanded={isExpanded} setIsExpanded={setIsExpanded} />
          {isExpanded ? (
            <div>
              <RawData data={value} displayCallback={displayCallback} />
            </div>
          ) : (
            <></>
          )}
        </dd>
      );
    }

    const displayValue =
      displayCallback === undefined ? value.toString() : displayCallback(key, value);

    return <dd>{displayValue}&nbsp;</dd>;
  };

  const renderElements = () => {
    if (data === undefined) {
      return [];
    }

    return Object.entries(data).map((keyValue) => {
      const [key, value] = keyValue;

      return (
        <div className={'bb-raw-data-key-value'} key={key}>
          <dt>{key}</dt>
          {renderDataElement(key, value)}
        </div>
      );
    });
  };

  return <dl className="dl-horizontal">{renderElements()}</dl>;
};
