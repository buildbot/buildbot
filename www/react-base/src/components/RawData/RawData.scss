.bb-raw-data-key-value {
  & > dt {
    float: left;
    width: 160px;
    clear: left;
    text-align: right;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
  }

  & > dd {
    text-align: left;
    padding-left: 20px;
    width: 70%;
    display: inline-block;
    box-sizing: border-box;
  }
}
