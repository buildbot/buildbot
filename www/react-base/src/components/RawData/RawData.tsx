/*
  This file is part of Buildbot.  Buildbot is free software: you can
  redistribute it and/or modify it under the terms of the GNU General Public
  License as published by the Free Software Foundation, version 2.

  This program is distributed in the hope that it will be useful, but WITHOUT
  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
  FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
  details.

  You should have received a copy of the GNU General Public License along with
  this program; if not, write to the Free Software Foundation, Inc., 51
  Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

  Copyright Buildbot Team Members
*/

import './RawData.scss';
import {useState} from "react";
import _ from "underscore";
import {isObservableArray, isObservableObject} from "mobx";
import {ArrowExpander} from "buildbot-ui";

const isArrayRaw = (v: any) => {
  return _.isArray(v) || isObservableArray(v);
}
const isObjectRaw = (v: any) => {
  return (_.isObject(v) || isObservableObject(v)) && !isArrayRaw(v);
}

const isArrayOfObjectsRaw = (v: any) => {
  return isArrayRaw(v) && v.length > 0 && isObjectRaw(v[0]);
}

type RawDataProps = {
  data: {[key: string]: any};
}

export const RawData = ({data}: RawDataProps) => {
  const [isExpanded, setIsExpanded] = useState(false);

  const renderArrayElements = (value: any[]) => {
    return (
      <ul>
        {value.map((v, index) => (
          <li key={index}>
            <RawData data={v}/>
          </li>
        ))}
      </ul>
    );
  }

  const renderDataElement = (value: any) => {
    if (value === null) {
      return <dd>{"null"}&nbsp;</dd>;
    }
    if (value === undefined) {
      return <dd>{"undefined"}&nbsp;</dd>;
    }
    if (isArrayOfObjectsRaw(value)) {
      return (
        <dd>
          <ArrowExpander isExpanded={isExpanded} setIsExpanded={setIsExpanded}/>
          {isExpanded ? renderArrayElements(value as any[]) : <span>{JSON.stringify(value)}</span>}
        </dd>
      );
    }
    if (isObjectRaw(value)) {
      return (
        <dd>
          <ArrowExpander isExpanded={isExpanded} setIsExpanded={setIsExpanded}/>
          {isExpanded ? <div><RawData data={value}/></div> : <></>}
        </dd>
      )
    }

    return <dd>{value.toString()}&nbsp;</dd>;
  }

  const renderElements = () => {
    if (data === undefined) {
      return [];
    }

    return Object.entries(data).map(keyValue => {
      const [key, value] = keyValue;

      return (
        <div className={'bb-raw-data-key-value'} key={key}>
          <dt>{key}</dt>
          {renderDataElement(value)}
        </div>
      );
    });
  }

  return (
    <dl className="dl-horizontal">
      {renderElements()}
    </dl>
  );
}
