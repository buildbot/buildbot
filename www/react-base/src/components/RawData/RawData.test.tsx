/*
  This file is part of Buildbot.  Buildbot is free software: you can
  redistribute it and/or modify it under the terms of the GNU General Public
  License as published by the Free Software Foundation, version 2.

  This program is distributed in the hope that it will be useful, but WITHOUT
  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
  FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
  details.

  You should have received a copy of the GNU General Public License along with
  this program; if not, write to the Free Software Foundation, Inc., 51
  Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

  Copyright Buildbot Team Members
*/

import renderer, {act} from 'react-test-renderer';
import {MemoryRouter} from "react-router-dom";
import {RawData} from './RawData';
import {observable} from "mobx";

function assertRenderSnapshot(data: {[key: string]: any}) {
  const component = renderer.create(
    <MemoryRouter>
      <RawData data={data}/>
    </MemoryRouter>
  );
  expect(component.toJSON()).toMatchSnapshot();
}

function assertRenderSnapshotExpanded(data: {[key: string]: any}) {
  const component = renderer.create(
    <MemoryRouter>
      <RawData data={data}/>
    </MemoryRouter>
  );
  act(() =>
    component.root.findByType("svg").props.onClick()
  );
  expect(component.toJSON()).toMatchSnapshot();
}

describe('RawData component', function() {
  it('simple object', () => {
    const obj = {
      str: "string",
      int: 123,
      float: 123.4,
      boolean: true,
      array: [321, "string", true],
      null: null,
      undefined: undefined,
    }

    assertRenderSnapshot(obj);
  });

  it('observable object', () => {
    const obj = observable({
      str: "string",
      int: 123,
      float: 123.4,
      boolean: true,
      array: observable([321, "string", true]),
      null: null,
      undefined: undefined,
    });

    assertRenderSnapshot(obj);
  });

  it('object with array of objects', () => {
    const obj = {
      array: [
        {str: "string"},
        {int: 123},
        {float: 123.4},
        {boolean: true},
        {null: null},
        {undefined: undefined},
      ],
    };

    assertRenderSnapshot(obj);
  });

  it('object with array of objects expanded', () => {
    const obj = {
      array: [
        {str: "string"},
        {int: 123},
        {float: 123.4},
        {boolean: true},
        {null: null},
        {undefined: undefined},
      ],
    };

    assertRenderSnapshotExpanded(obj);
  });

  it('observable object with array of objects', () => {
    const obj = observable({
      array: [
        observable({str: "string"}),
        observable({int: 123}),
        observable({float: 123.4}),
        observable({boolean: true}),
        observable({null: null}),
        observable({undefined: undefined}),
      ],
    });

    assertRenderSnapshot(obj);
  });
});
