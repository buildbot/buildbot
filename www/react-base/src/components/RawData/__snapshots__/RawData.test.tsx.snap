// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RawData component object with array of objects 1`] = `
<dl
  className="dl-horizontal"
>
  <div
    className="bb-raw-data-key-value"
  >
    <dt>
      array
    </dt>
    <dd>
      <svg
        className="rotate clickable"
        fill="currentColor"
        height="1em"
        onClick={[Function]}
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M256 8c137 0 248 111 248 248S393 504 256 504 8 393 8 256 119 8 256 8zm113.9 231L234.4 103.5c-9.4-9.4-24.6-9.4-33.9 0l-17 17c-9.4 9.4-9.4 24.6 0 33.9L285.1 256 183.5 357.6c-9.4 9.4-9.4 24.6 0 33.9l17 17c9.4 9.4 24.6 9.4 33.9 0L369.9 273c9.4-9.4 9.4-24.6 0-34z"
        />
      </svg>
      <span>
        [{"str":"string"},{"int":123},{"float":123.4},{"boolean":true},{"null":null},{}]
      </span>
    </dd>
  </div>
</dl>
`;

exports[`RawData component object with array of objects expanded 1`] = `
<dl
  className="dl-horizontal"
>
  <div
    className="bb-raw-data-key-value"
  >
    <dt>
      array
    </dt>
    <dd>
      <svg
        className="rotate clickable"
        fill="currentColor"
        height="1em"
        onClick={[Function]}
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M504 256c0 137-111 248-248 248S8 393 8 256 119 8 256 8s248 111 248 248zM273 369.9l135.5-135.5c9.4-9.4 9.4-24.6 0-33.9l-17-17c-9.4-9.4-24.6-9.4-33.9 0L256 285.1 154.4 183.5c-9.4-9.4-24.6-9.4-33.9 0l-17 17c-9.4 9.4-9.4 24.6 0 33.9L239 369.9c9.4 9.4 24.6 9.4 34 0z"
        />
      </svg>
      <ul>
        <li>
          <dl
            className="dl-horizontal"
          >
            <div
              className="bb-raw-data-key-value"
            >
              <dt>
                str
              </dt>
              <dd>
                string
                 
              </dd>
            </div>
          </dl>
        </li>
        <li>
          <dl
            className="dl-horizontal"
          >
            <div
              className="bb-raw-data-key-value"
            >
              <dt>
                int
              </dt>
              <dd>
                123
                 
              </dd>
            </div>
          </dl>
        </li>
        <li>
          <dl
            className="dl-horizontal"
          >
            <div
              className="bb-raw-data-key-value"
            >
              <dt>
                float
              </dt>
              <dd>
                123.4
                 
              </dd>
            </div>
          </dl>
        </li>
        <li>
          <dl
            className="dl-horizontal"
          >
            <div
              className="bb-raw-data-key-value"
            >
              <dt>
                boolean
              </dt>
              <dd>
                true
                 
              </dd>
            </div>
          </dl>
        </li>
        <li>
          <dl
            className="dl-horizontal"
          >
            <div
              className="bb-raw-data-key-value"
            >
              <dt>
                null
              </dt>
              <dd>
                null
                 
              </dd>
            </div>
          </dl>
        </li>
        <li>
          <dl
            className="dl-horizontal"
          >
            <div
              className="bb-raw-data-key-value"
            >
              <dt>
                undefined
              </dt>
              <dd>
                undefined
                 
              </dd>
            </div>
          </dl>
        </li>
      </ul>
    </dd>
  </div>
</dl>
`;

exports[`RawData component observable object 1`] = `
<dl
  className="dl-horizontal"
>
  <div
    className="bb-raw-data-key-value"
  >
    <dt>
      str
    </dt>
    <dd>
      string
       
    </dd>
  </div>
  <div
    className="bb-raw-data-key-value"
  >
    <dt>
      int
    </dt>
    <dd>
      123
       
    </dd>
  </div>
  <div
    className="bb-raw-data-key-value"
  >
    <dt>
      float
    </dt>
    <dd>
      123.4
       
    </dd>
  </div>
  <div
    className="bb-raw-data-key-value"
  >
    <dt>
      boolean
    </dt>
    <dd>
      true
       
    </dd>
  </div>
  <div
    className="bb-raw-data-key-value"
  >
    <dt>
      array
    </dt>
    <dd>
      321,string,true
       
    </dd>
  </div>
  <div
    className="bb-raw-data-key-value"
  >
    <dt>
      null
    </dt>
    <dd>
      null
       
    </dd>
  </div>
  <div
    className="bb-raw-data-key-value"
  >
    <dt>
      undefined
    </dt>
    <dd>
      undefined
       
    </dd>
  </div>
</dl>
`;

exports[`RawData component observable object with array of objects 1`] = `
<dl
  className="dl-horizontal"
>
  <div
    className="bb-raw-data-key-value"
  >
    <dt>
      array
    </dt>
    <dd>
      <svg
        className="rotate clickable"
        fill="currentColor"
        height="1em"
        onClick={[Function]}
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M256 8c137 0 248 111 248 248S393 504 256 504 8 393 8 256 119 8 256 8zm113.9 231L234.4 103.5c-9.4-9.4-24.6-9.4-33.9 0l-17 17c-9.4 9.4-9.4 24.6 0 33.9L285.1 256 183.5 357.6c-9.4 9.4-9.4 24.6 0 33.9l17 17c9.4 9.4 24.6 9.4 33.9 0L369.9 273c9.4-9.4 9.4-24.6 0-34z"
        />
      </svg>
      <span>
        [{"str":"string"},{"int":123},{"float":123.4},{"boolean":true},{"null":null},{}]
      </span>
    </dd>
  </div>
</dl>
`;

exports[`RawData component simple object 1`] = `
<dl
  className="dl-horizontal"
>
  <div
    className="bb-raw-data-key-value"
  >
    <dt>
      str
    </dt>
    <dd>
      string
       
    </dd>
  </div>
  <div
    className="bb-raw-data-key-value"
  >
    <dt>
      int
    </dt>
    <dd>
      123
       
    </dd>
  </div>
  <div
    className="bb-raw-data-key-value"
  >
    <dt>
      float
    </dt>
    <dd>
      123.4
       
    </dd>
  </div>
  <div
    className="bb-raw-data-key-value"
  >
    <dt>
      boolean
    </dt>
    <dd>
      true
       
    </dd>
  </div>
  <div
    className="bb-raw-data-key-value"
  >
    <dt>
      array
    </dt>
    <dd>
      321,string,true
       
    </dd>
  </div>
  <div
    className="bb-raw-data-key-value"
  >
    <dt>
      null
    </dt>
    <dd>
      null
       
    </dd>
  </div>
  <div
    className="bb-raw-data-key-value"
  >
    <dt>
      undefined
    </dt>
    <dd>
      undefined
       
    </dd>
  </div>
</dl>
`;
